reps/2024-04-24-ray-obs-support.md

```diff
@@ -0,0 +1,103 @@
+
+- Third, after submitting the ray job, the ray cluster automatically downloads, uncompresses, and executes the codes from the OBS
+
+![Figure 1](2024-04-24-ray-obs-support/fig1.png)
+
+
+### Should this change be within `ray` or outside?
+
+main `ray` project. Changes are made to Ray core components.
+
+## Stewardship
+
+### Required Reviewers
+
+- @jjyao
+- @ericl
+
+
+### Shepherd of the Proposal (should be a senior committer)
+
+- @ericl
+
+
+### Design and Architecture
+
+To submit a Ray job through OBS, perform the following steps:
+
+1. Compress the code into a zip or jar package and put it in OBS
+
+2. You can use environment variables or configuration files to access the AK, SK, and Endpoint of OBS
+
+3. When you submit a ray job, specify the path of the OBS service
+
+```python
+job_id = client.submit_job(
+    entrypoint="python script.py",
+    runtime_env={"working_dir": "obs://example_bucket/example_file.zip"}
+)
+```
+
+4. The ray cluster automatically downloads the example_file.zip from the specified OBS bucket, decompresses it, and then runs the entry file script.py in the working path
+
+
+## Design Insights
+
+Modify the source code of Ray so that it can download and run the OBS code as follows:
+
+1. Parse the OBS path, for example, "obs://example_bucket/example_file.zip";
+
+2. Read the AK, SK, and endpoint of OBS through environment variables and configuration files to access the remote OBS path.
+
+3. Download the example_file.zip from the specified OBS bucket, decompress it, and execute the user's code
+
+We can extend the ray project to support obs protocol, enabling the ray cluster to parse the obs URI and download the codes from obs:
+
+1. parse obs URI, such as "obs://example_bucket/example_file.zip";
+
+2. config the obs AK, SK, and Endpoint via environment variables and config files;
+
+3. the ray cluster automatically downloads and execute the codes from obs.
+
+## Implementation Analysis
+
+To extend the ray project for obs, we first need to figure out the workflow of parsing and accessing remote URIs, which is shown in Figure 2.
+![Figure 2](2024-04-24-ray-obs-support/fig2.png)
+
+
+After the user submitting remote ray jobs, the ray cluster calls download_and_unpack_package function to download and uncrompress the remote files, as shown in Figure 2. To extend the ray project for OBS, we should extend the download_and_unpack_package function to support the OBS scheme, which is implemented via the following two steps.
+
+**Step 1**: Extend the **parse_uri** function to parse the OBS URIs, which is implemented in the file [ray/_private/runtime_env/packaging](https://github.com/ray-project/ray/blob/master/python/ray/_private/runtime_env/packaging.py).
+
+**Step 2**: Extend the third-party library [smart_open](https://github.com/piskvorky/smart_open) to read OBS objects, which is suggested to implement 3 interfaces, i.e., parse uri, open_uri, and open, as shown in Extending smart_open.
```

Comment:
So you will create a PR in smart_open to support OBS first?

Reply:
Yes, firstly I implement required logic in smart_open.

Reply:
Are you planning to upstream this functionality to `smart_open` or extend it within the Ray codebase?

Reply:
firstly, I will add functionality to smart_open, then I will add logic to ray packaging.py